New search

RJ722 opened the pull request:
## Summary of the changes in this pull request

Previously, a search would only return results with exact matches, that is if the search string is a substring of the text in the window buffer.  With this new edition, we support searching for multiple words. A result would only be shown if *all* of those words appear at least once in the text.

~~There is a bug in there due to which highlights in the main windows aren't working properly. I have figured that one needs to have the day opened on the side, go to the search bar, and hit enter again for the highlights to work correctly. I have a hunch that this is because I've skipped the "scroll to query" part -- which I'm guessing refreshes the view? Just a hunch, I can be totally off on this.~~

EDIT: Whenever a search result is clicked, `Day.show_day` method is called to update the view. It used the `self.search_text` to take care of highlighting and scrolling to search results. I updated it to use the new `self.search_queries` instead, and everything appears to be working! \o/

## Pull request checklist
<!--- Go over the following points, and put an `x` into all boxes that apply. -->

- [x] I have added an entry in `CHANGELOG.md` including my name and issue and/or pull request number.
- [x] If applicable: I have removed the corresponding entry in `TODO.md`.

jendrikseipp commented on rednotebook/gui/editor.py:
```diff
@@ -131,8 +131,8 @@ def highlight(self, text):
-        if text:
-            for match_start, match_end in self.iter_search_matches(text):
+        for query in enumerate(queries):
+            for match_start, match_end in self.iter_search_matches(query):
```
`query` is now a tuple, e.g., `(0, "cat")`. Doesn't that break the call to iter_search_matches? To me it seems the `enumerate` should be removed.

RJ722 replied:
Oops, you're right. I added an `enumerate` when I was testing something and forgot to remove this. On it now.

jendrikseipp commented on rednotebook/gui/search.py:
```diff
@@ -60,18 +60,21 @@ def search(self, search_text):
+            # Would it make sense to inspect the data returned by
+            # `self.journal.search` -- which will have information on what
+            # query matched fist, and it's index?
+            pass
```
I think it's fine to use the old code for this. It should be fast enough. We can simply scroll to the first query that's not part of the day's date. For this we can pass `tags` and queries` to `scroll_to_text` and let it pick the first non-date query.

RJ722 replied:
> For this we can pass `tags` and `queries` to `scroll_to_text` and let it pick the first non-date query.

Hey, not sure what you meant by that: Do we need to change the implementation of `scroll_to_text` to accept a list type argument (where we'd pass `tags + queries`) instead of `text` (string type) -- Because, my understanding is that `it.forward_search` (which is used internally by `scroll_to_text`) only accepts string arguments.

Even if we do that, it won't be able to decide which query to pick automatically, right?

jendrikseipp replied:
Yes, we need to change `scroll_to_text`. It gets the information about the whole query and then loops over the info and chooses the first string that is not part of the day's date (e.g., "2020-02-28"). I think the query information can be passed as a list of strings (by concatenating tags and queries).

RJ722 replied:
Got it, thanks for the prompt reply!

I am trying to figure out if there's a way to access the current day's string, from `scroll_to_text` to no avail. I wanted to ask you -- is there an easy way to refer to the `Day` object from `search_to_text` method?

RJ722 replied:
Ah, nvm, I figured it -- it was as simple as calling `self.day`. 🤦

jendrikseipp reviewed (changes requested):
Looks good :-) I only have a few comments. Maybe the bug you're seeing will be gone after taking care of these.

RJ722 commented:
@jendrikseipp After the latest changes, indeed the bug went away. (I was actually still checking the old `search_text`, whereas I had renamed it to `search_queries`.)

I've found another bug now which causes an exception if only a date string is passed to search (easily fixable). I'll do that. Plus, I was thinking that I should change all occurrences of `queries` to `words`. What do you think?

jendrikseipp commented:
> Plus, I was thinking that I should change all occurrences of `queries` to `words`. What do you think?

Sounds good!

RJ722 commented:
Took me a lot longer, but pheww. finally done. :D

jendrikseipp commented:
Awesome! I hope to be able to review this soon.

jendrikseipp commented on rednotebook/gui/browser.py:
```diff
@@ -65,12 +65,13 @@ def set_font_size(self, size):
-                self.search_text, WebKit2.FindOptions.CASE_INSENSITIVE, MAX_HITS
-            )
+            self.search_words = search_words
+            for query in search_words:
```
query --> word

jendrikseipp commented on rednotebook/gui/browser.py:
```diff
@@ -65,12 +65,13 @@ def set_font_size(self, size):
+            for query in search_words:
+                self.get_find_controller().search(
+                    query, WebKit2.FindOptions.CASE_INSENSITIVE, MAX_HITS
+                )
```
Can we stop after the first word is found?

RJ722 replied:
A little confused here since [the docs](https://lazka.github.io/pgi-docs/WebKit2-4.0/classes/FindController.html#WebKit2.FindController.search) say that the `search` method performs the search asynchronously. If so, would this be possible?

jendrikseipp commented on rednotebook/gui/main_window.py:
```diff
@@ -814,17 +814,29 @@ def _get_buffer(self, key, text):
+    def scroll_to_text(self, words):
+        """
+        Finds the first non-date word in words, and passes it on to
+        `Editor.scroll_to_text`.
```
Method docs should use imperative: finds --> find, passes --> pass

jendrikseipp commented on rednotebook/gui/search.py:
```diff
@@ -52,26 +52,21 @@ def on_entry_activated(self, entry):
-
-        # Highlight all occurences in the current day's text
-        self.main_window.highlight_text(search_text)
+        # Highlight all occurrences in the current day's text
```
Finish sentences with a dot.

jendrikseipp commented on rednotebook/data.py:
```diff
@@ -200,41 +200,54 @@ def search(self, text, tags):
+                if word in str(self):
+                    # We don't want to show search results from the text matching
+                    # this date.
+                    words.remove(word)
```
list.remove() is dangerous since it uses a linear search in the list. It's better to create a new empty list and add all the non-date words to it. This can be done in a list comprehension: 
``` python
non_date_words = [word for word in words if word not in str(self)]
```

And if word contains "2021" and "2021-04", we only want to add it to the results once. So we can do 

```python
words_contain_date = len(non_date_words) != len(words)
if words_contain_date:
    results.append(get_text_with_dots(self.text, 0, TEXT_RESULT_LENGTH))
```

RJ722 replied:
Nice catch Jendrik!

jendrikseipp commented on rednotebook/data.py:
```diff
@@ -200,41 +200,54 @@ def search(self, text, tags):
+                    # this date.
+                    words.remove(word)
+                    results.append(get_text_with_dots(self.text, 0, TEXT_RESULT_LENGTH))
+            # if all the words matched agains the date, return
```
`If all the words matched against the date, return.`

jendrikseipp commented on rednotebook/data.py:
```diff
@@ -200,41 +200,54 @@ def search(self, text, tags):
+                    results.append(get_text_with_dots(self.text, 0, TEXT_RESULT_LENGTH))
+            # if all the words matched agains the date, return
+            if not words:
+                return str(self), words
```
words --> results

jendrikseipp commented on rednotebook/data.py:
```diff
@@ -200,41 +200,54 @@ def search(self, text, tags):
-        result_text = get_text_with_dots(
-            self.text, occurence, occurence + len(search_text), found_text
+    def search_in_text(self, words):
+        """All words should be present in the text"""
```
`If all words are in the text, return a suitable text substring. Otherwise, return None.`

jendrikseipp commented on rednotebook/data.py:
```diff
@@ -200,41 +200,54 @@ def search(self, text, tags):
+        """All words should be present in the text"""
+        matches = {}
+        for word in words:
+            index = self.text.upper().find(word.upper())
```
Let's use lower() instead of upper().

jendrikseipp commented on rednotebook/data.py:
```diff
@@ -200,41 +200,54 @@ def search(self, text, tags):
-            self.text, occurence, occurence + len(search_text), found_text
+    def search_in_text(self, words):
+        """All words should be present in the text"""
+        matches = {}
```
No need for a dict. Just update the smallest position (initialized to sys.maxsize) while looping over the words.

RJ722 replied:
Yep, this simplified the code. Thanks!

jendrikseipp reviewed (changes requested):
I finally found the time to have a look :-) This is shaping up nicely!

RJ722 commented:
Hey, thanks for the review Jendrik! \o/

I believe I've addressed all of the comments. I also took the liberty and renamed the `DayEditor.scroll_to_text` to `DayEditor.scroll_to_non_date_text` which will help prevent the confusion because of the method in the parent class by the same name. Let me know if you don't like that and would like to revert.

RJ722 commented:
Closing and re-opening to trigger CI again.

jendrikseipp reviewed (changes requested):
Thanks for taking care of my comments! I looked into the webkit highlighting. Now everything should work. I only have minor comments remaining.

jendrikseipp commented:
@RJ722 Do you still want to pursue this? If so, it seems you'll have to merge with master first and then take care of my latest few comments.
